runtime/src/comm/ofi/comm-ofi-oob-pmi2.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// PMI2-based out-of-band support for the OFI-based Chapel comm layer.
+//
+
+#include "chplrt.h"
+#include "chpl-env-gen.h"
+
+#include "chpl-comm.h"
+#include "chpl-mem.h"
+#include "chpl-mem-sys.h"
+#include "chpl-gen-includes.h"
+#include "chplsys.h"
+#include "error.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+#include "comm-ofi-internal.h"
+
+
+#define PMI2_SUCCESS 0
+#define PMI2_ID_NULL -1
+
+extern int PMI2_Initialized(void);
+extern int PMI2_Init(int* spawned, int* size, int* rank, int* appnum);
+extern int PMI2_Finalize(void);
+extern int PMI2_KVS_Put(const char key[], const char value[]);
+extern int PMI2_KVS_Fence(void);
+extern int PMI2_KVS_Get(const char* jobid, int src_pmi_id,
+                        const char key[], char value[],
+                        int maxvalue, int* vallen);
+
```

Comment:
Missed this initially, but can you explain why we need to extern? Is there no header we can reliably include?

Reply:
That's it exactly -- there's not necessarily a header.  And if there is a header, it might actually be an MPI one or, more specifically, an MPICH one.  (The PMI and newer PMI-2 interfaces were both defined by MPICH2; they're not standalone although it can look that way on Cray XC etc.)  I'll add a comment to this effect.